[WIP] Adding ARCore

BastiaanOlij opened the pull request:
This is still a Work In Progress but we're making good progress. Please feel free to try and give feedback.

Right now it almost runs, it downloads the libraries, requests permissions and has all the code that should allow us to do positional tracking. No camera image yet but we just have one shader issue to fix. 

For those who want to help out you can best clone the ARCore branch on my fork and just compile it, 
I'm maintaining a test project here:
https://github.com/BastiaanOlij/godot3_test_projects/tree/ARCore/ARCore

Or create a project with the following code in your main script:
```
func _ready():
	# ARCore initialisation should be in _ready
	var arcore = ARVRServer.find_interface('ARCore')
	if arcore and arcore.initialize():
		# this just means we started our initialisation process successfully
		get_viewport().arvr = true
			
		# assign our camera
		get_viewport().get_camera().environment.background_camera_feed_id = arcore.get_camera_feed_id()
```
Also make sure your environment is setup to use a background camera

Todos:
- fix shader issues around external sampler
- fix matrix issues
- add support for detecting anchors
- add code for handling when Godot needs to be pauzed (reacting to notifications)

BastiaanOlij commented:
One thing I could use some help with is including ARCore libraries itself. I've reserved a place for these within  thirdparty/arcore, they are originally sourced from here: https://github.com/google-ar/arcore-android-sdk

Right now you need to download the Android SDK, compile the hello ar example, this will cause gradlew to extract a number of support libraries, and copy those support libraries into thirdparty/arcore/jni

The license around these .so files may not be compatible with MIT so I don't know if we can just include them in the thirdparty/arcore folder. If not we may need to look into scons or gradlew downloading them as components somehow.

The other issue I'm trying to solve is that while we're compiling these .so files into our project (see the detect.py changes) Android is still complaining that the files can't be loaded. It looks like they have dynamic counter parts.

aaronfranke commented:
Is this only for Android, or does it also work with other AR platforms like Windows MR?

HeadClot commented:
> Is this only for Android, or does it also work with other AR platforms like Windows MR?

It is only for android as far as I know.

BastiaanOlij commented:
Only for Android, there is a separate PR for ARKit for iOS which is much
further along.

MR is an entirely different story. Specifically for hololens you have to
build UWP applications that only support DirectX. Godot uses Angle to
create an OpenGL to directx bridge. There is a special version for
microsofts holographic API but last time i played around with it I couldn’t
get it to work

On Sun, 24 Feb 2019 at 4:47 pm, Aaron Franke <notifications@github.com>
wrote:

> Is this only for Android, or does it also work with other AR platforms
> like Windows MR?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/26221#issuecomment-466736632>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2vaUmePKUTemqKHvGePOw6ZXRFeTeAks5vQid3gaJpZM4bOVlp>
> .
>
-- 
Kindest regards,

Bastiaan Olij

https://www.facebook.com/bastiaan.olij
https://twitter.com/mux213
https://www.youtube.com/BastiaanOlij
<https://www.youtube.com/channel/UCrbLJYzJjDf2p-vJC011lYw>
https://github.com/BastiaanOlij

BastiaanOlij commented:
Ok, it now pulls in the AR libraries from Google through gradlew. Stores them in thirdparty/godot, then copies the files it needs into platform/android/java/jni

Now compiles and starts up without crashing, but on my pocophone it fails on ArCoreApk_requestInstall
I have working ARCore apps so I'm guessing it's a rights issue. 

Anyway, getting closer to a version that does tracking :)

BastiaanOlij commented on platform/android/SCsub:
```diff
@@ -169,3 +169,15 @@ if lib_arch_dir != '':
         stl_lib_path = str(env['ANDROID_NDK_ROOT']) + '/sources/cxx-stl/llvm-libc++/libs/' + lib_arch_dir + '/libc++_shared.so'
         env_android.Command(out_dir + '/libc++_shared.so', stl_lib_path, Copy("$TARGET", "$SOURCE"))
+
+    if 'module_arcore_enabled' in env and env['module_arcore_enabled']:
```
Need to see if we can move this into modules\arcore\SCSub somehow

akien-mga replied:
Fixed in BastiaanOlij/godot#9.

BastiaanOlij commented on platform/android/detect.py:
```diff
@@ -298,6 +298,17 @@ def mySpawn(sh, escape, cmd, args, env):
     env.Append(CPPFLAGS=['-DANDROID_ENABLED', '-DUNIX_ENABLED', '-DNO_FCNTL'])
     env.Append(LIBS=['OpenSLES', 'EGL', 'GLESv3', 'android', 'log', 'z', 'dl'])
 
+    ## arcore, should see about moving this into our module
```
Same with this, need to see if we can move it into modules\arcore somehow, not sure if its possible

akien-mga replied:
Fixed in BastiaanOlij/godot#9.

BastiaanOlij commented on platform/android/AndroidManifest.xml.template:
```diff
@@ -12,10 +12,12 @@
                       android:xlargeScreens="true"/>
 
     <uses-feature android:glEsVersion="0x00020000" android:required="true" />
+    <uses-feature android:name="android.hardware.camera.ar" android:required="true"/>
```
Features should become part of our export settings

BastiaanOlij commented on platform/android/AndroidManifest.xml.template:
```diff
@@ -12,10 +12,12 @@
 $$ADD_PERMISSION_CHUNKS$$
 
     <application android:label="@string/godot_project_name_string" android:icon="@drawable/icon" android:allowBackup="false" tools:ignore="GoogleAppIndexingWarning" $$ADD_APPATTRIBUTE_CHUNKS$$ >
+        <meta-data android:name="com.google.ar.core" android:value="required" />
```
This too needs to be redone as an export setting

BastiaanOlij commented:
I've added the AndroidPermissions singleton code in this for now, with this and some GDScript code the camera permissions now work properly. This needs to be redone and properly implemented outside of this PR (see #26553 and #26506 for related discussions) so it's just in there temporarily.

The proper implementation here would be for the ARCore module to interact with the permissions module, ask it if camera permissions have been granted and request permissions if they have not. This way there is no code needed on the GDScript side.

ARCore still fails to load with an non-descriptive error. I think the problem lies at the start of _resume where I'm retrieving the activity.

whplh commented:
Hey there. To keep it concise. I'm using a debug build of this commit with the ARCore sample project from BastiaanOlij :+1: :
- From the template manager I get the message: "Error getting the list of mirrors"
    - Template manger in 3.06 works fine
- I manually installed/tried the export templates beta1 and beta11 from https://downloads.tuxfamily.org/godotengine/3.1/ 
- When I try to export the sample project to Android I get: 

    > Failed to export the project for platform 'Android' Export templates seem to be missing or invalid
    - Other export platforms work
    - I can export with Godot 3.0.6 to Android -no problems

**Which export templates should I use?**

Note: I could not find ARCore's license in Godots UI : Help->About->Thirdparty License  or is it "The Android Open Source Project" ?

Anyway - Great work! Any hints on how to resolve my issues would be great. Happy weekend.

akien-mga commented:
> Which export templates should I use?

You should build export templates yourself from the same commit you used to build the editor. There are no export templates prebuilt for that commit, which is why you can't use the export template manager.
The simplest is to [build your own APK](http://docs.godotengine.org/en/latest/development/compiling/compiling_for_android.html) and add it as Custom Debug Package in the Android preset.

> Note: I could not find ARCore's license in Godots UI : Help->About->Thirdparty License or is it "The Android Open Source Project" ?

Those licenses are extracted from the `COPYRIGHT.txt` file. ARCore is not in there yet, and being a module it should likely not be hardcoded but added conditionally - which is not possible at the moment, see #23283.

BastiaanOlij commented:
@whplh what akien said :)
Also this currently is far from ready for any testing, it doesn't run yet. Still trying to figure out a bunch of build issues.

whplh commented:
Just in case some else has the same problem as me:
running:
```
gradlew clean 

```
Would probably have resolved the problem of the missing manifest after switching architectures.

BastiaanOlij commented:
Ok, this is starting to do something.

First, notifications aren't being sent to ARVRInterfaces yet, so when you give camera permission we don't get our signal to do our next step. Give camera permission and restart the app.
Then, request install doesn't work yet, the code is commented out for now. Make sure ARCore is installed on your phone!
Finally compiling has proven to have some interesting issues. The following works for me on windows:
```
scons platform=android target=release_debug -j8
scons platform=android target=release_debug android_arch=arm64v8 -j8
cd platform\android\java
gradlew clean
gradlew build
```
This builds the debug build ONLY, if you need the release build add that in too.
The gradlew clean is very important

reduz commented:
can't wait to merge once it's working!

BastiaanOlij reviewed (commented):
> can't wait to merge once it's working!

Still a lot to do :) I'll round off that restructure PR first and then I need to add support for an Android specific texture type to the visual server. And then finish the camera server so we can actually display the camera image, and then I can finally start making this do something that makes sense :)

BastiaanOlij commented:
Ok, just rebased and fixed a few small things. The initialization of ARCore now works properly. The view and projection matrices are still wrong, I think its got to do with providing the incorrect orientation data to ARCore so it will only be correct if you hold your phone in one orientation. Not 100% sure.

I'm going to work on the camera next.

BastiaanOlij commented:
Nearly got the camera server to work with ARCore, getting late so checked in all the changes so far.

Note that:
- we're not getting notifications yet so we don't react on gaining camera permissions. You need to restart the app after providing camera permissions before tracking will work
- background image is not being shown yet, nearly there
- the camera is reversed, so we're moving the wrong way, need to rotate it 180 degrees, haven't had time to look at that yet.
- orientation has not yet been fixed. 

I'm also going to move some changes into the camera server PR and move a few of the visual server changes into its own PR

BastiaanOlij committed:
Adding ARCore interface

BastiaanOlij commented:
Ok, work for today was mostly restructuring after putting a bunch of changes into the PRs they belong in. 

I'm nearly at a point where the camera image renders. In order to access the camera images we need to use a special extention on Android. Turns out this also needs a special sampler in the shader called `samplerExternalOES`, I've rigged it in for now. 

Problem is that that results in a compilation error:
`ERROR: 0:48: 'samplerExternalOES' : requires extension GL_OES_EGL_image_external_essl3 to be enabled`

@reduz any ideas how I enable that extension?

GeorgeS2019 commented:
Anyone has an updated view of Godot future support for Hololens 2? The Hololens 2 emulator has just being released!

BastiaanOlij commented:
Have to look into it, hololens 1 runs UWP apps only, that means DirectX
only with OpenGL support through ANGLE but the changes needed to this for
hololens don’t seem to be maintained. If hololens 2 is still UWP that may
pose an issue. Microsoft is putting their weight under OpenXR however which
may solve this issue.

ARCore however is google/android only so wrong place to discuss this:)

On Sat, 20 Apr 2019 at 10:44 pm, GeorgeS2019 <notifications@github.com>
wrote:

> Anyone has an updated view of Godot future support for Hololens 2? The
> Hololens 2 emulator has just being released!
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/26221#issuecomment-485111764>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO262OMKSFXNDVIY6KCYELPRMFZVANCNFSM4GZZLFUQ>
> .
>
-- 
Kindest regards,

Bastiaan Olij

https://www.facebook.com/bastiaan.olij
https://twitter.com/mux213
https://www.youtube.com/BastiaanOlij
<https://www.youtube.com/channel/UCrbLJYzJjDf2p-vJC011lYw>
https://github.com/BastiaanOlij

GeorgeS2019 commented:
@BastiaanOlij 

I am new to Godot! However I see so many potential in it as a replacement for Unity4Hololens

Could u suggest how to raise community interest on Godot4Hololens2 ?

Hololens2 runs on Snapdragon 850, which means a variant of Windows 10 on ARM  => which means Godot needs to compile on Win10 on ARM.

Are you aware if the community have discussed on this?

BastiaanOlij commented:
Not really aware of a lot of discussion but keep in mind, most of us don’t
have access to the hardware so unless Microsoft steps in or a university
gives a student access to the hardware it’ll probably not happen any time
soon. There is one thing that may happen but until thats more concrete i
can’t say much about it.

I did get an MR headset with which i hope to do more work on OpenXR but got
stuck on native support due to the issues around UWP and OpenGL. If it
wasn’t for that godot would be a smashing fit for hololens

Godot is already being compiled for ARM for other platforms so i doubt that
will be an issue

On Sun, 21 Apr 2019 at 8:31 am, GeorgeS2019 <notifications@github.com>
wrote:

> @BastiaanOlij <https://github.com/BastiaanOlij>
>
> I am new to Godot! However I see so many potential in it as a replacement
> for Unity4Hololens
>
> Could u suggest how to raise community interest on Godot4Hololens2 ?
>
> Hololens2 runs on Snapdragon 850, which means a variant of Windows 10 on
> ARM => which means Godot needs to compile on Win10 on ARM.
>
> Are you aware if the community have discussed on this?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/26221#issuecomment-485186600>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO262JP65BBRKFMKBGQR7LPROKUXANCNFSM4GZZLFUQ>
> .
>
-- 
Kindest regards,

Bastiaan Olij

https://www.facebook.com/bastiaan.olij
https://twitter.com/mux213
https://www.youtube.com/BastiaanOlij
<https://www.youtube.com/channel/UCrbLJYzJjDf2p-vJC011lYw>
https://github.com/BastiaanOlij

BastiaanOlij commented:
Notifications now work which means after camera permissions are accepted ARCore will attempt to reinitialize and if at the next question ARCore needs to be installed it'll attempt reinitialize after it has been installed.

Also changed the view matrices but there is still something not right here. I think this may be related to orientation issues.

BastiaanOlij commented:
Alright, major update and breakthrough on this project today. There are a few hacks in there that need to be solved better (mainly the way ARCore needs extensions to render the camera background) and I have to move a few things into the appropriate PRs this is based on. However this now does all the important things.

Camera is rendered on the background, tracking is working properly now, only big thing left to do is detecting surfaces and we'll have parity with the ARKit build.

GeorgeS2019 commented:
FYI: https://github.com/microsoft/microsoft-ui-xaml/issues/686

GeorgeS2019 commented:
@BastiaanOlij   could you provide some technical suggestion how Godot can be incorporated into Windows UI Control.   There is currently no 3D viewer/game engine control of similar features as godot. I believe there is role for Godot to play in addressing the gap.

BastiaanOlij commented:
@GeorgeS2019 the biggest hurdle I've experienced so far is Microsofts reliance on UWP and DirectX only architecture. There is no OpenGL support so OpenGL based applications like Godot require to use a conversion library called Angle. There is a special version of Angle to support MS Holographic which has not had any fixes applied in the last 3 years and fails to compile on newer versions of Visual Studio.

Microsoft has put their weight behind OpenXR and I believe they are receptive to Vulkan support, I don't know the status of these in relation to Hololens 2 but those may open the door to having Godot run on Hololens.

Then taking the next step to create a 3D UI, I think people need to start experimenting with this. I've not dived into it deep enough yet so it's hard to say from my point of view what best practices are and what the options here are, but any game engine IMHO is ideally placed to form a basis here.

GeorgeS2019 commented:
@BastiaanOlij  this is an [example](https://github.com/mattleibow/ANGLE-Minimal) using ANGLE (OpenGL) in XAML . MS Holographics focus is now on [WebGL ](https://github.com/Microsoft/HoloJS), which depends on [WebGL in UWP](http://holojs.azurewebsites.net/help/uwp-apps.html)

I am lobby for Godot because I know the conversion from Blender to Godot is more progressive than what can be acheived through these HoloJS 3D environment.

GeorgeS2019 commented:
@BastiaanOlij   FYI [Godot is under evaluation](https://github.com/godotengine/godot/issues/28830#issuecomment-491993462)

BastiaanOlij commented:
Rebased, merged in the external texture changes and applies GLES2 changes. Still testing, but compiling is taking longer then I'm staying awake :)

BastiaanOlij commented:
Rebased this but only tested as far as compiling goes. Also clang-format was being difficult so I will likely have to fix one or two things up.

I'll do proper testing soon. There is still some work left to do around making plane detection work.

BastiaanOlij commented on drivers/gles2/rasterizer_scene_gles2.cpp:
```diff
@@ -2991,7 +3001,7 @@ void RasterizerSceneGLES2::render_scene(const Transform &p_cam_transform, const
 
 		if (storage->frame.current_rt->multisample_active) {
-			// Resolve framebuffer to front buffer before copying
+		// Resolve framebuffer to front buffer before copying
```
@akien-mga any idea why clang-format is flipping this and later on line 3063 as well? I'll fix it up when I have time to work on this next.

akien-mga replied:
clang-format was always a bit difficult around preprocessor defines, but in recent versions it seems to work ok.

What version do you use?

BastiaanOlij replied:
Still on 5.0.0 on Windows

akien-mga replied:
That's too old, and indeed one of the versions which was buggy around `#define` and `#ifdef`.

We require 8.0.0 or later now: http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#using-clang-format-locally

BastiaanOlij replied:
Downloading 5.0.2, or should I go to a higher version?

BastiaanOlij replied:
ok 8 it is :)

akien-mga replied:
Too old too, you need 8.0.0 or later.

BastiaanOlij replied:
Same issue with 8. When it's wrong it will correct it, when its correct it'll make it wrong. But atleast with a commit I can override it, it wouldn't let me when cherry-picking :)

BastiaanOlij commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -4492,6 +4501,7 @@ void RasterizerSceneGLES3::render_scene(const Transform &p_cam_transform, const
 				} else {
 					// don't have a feed, just show greenscreen :)
 					clear_color = Color(0.0, 1.0, 0.0, 1.0);
+					storage->frame.clear_request = true;
```
This may have been added back in due to this project going on for awhile, not sure if this should be here

akien-mga replied:
If you had a merge conflict, it would likely have been with 02ea99129e8f3882914431374c60a7d80c5146e1 where this line was not added.

BastiaanOlij commented on platform/android/java_godot_wrapper.cpp:
```diff
@@ -95,6 +97,15 @@ jobject GodotJavaWrapper::get_class_loader() {
 }
 
+jobject GodotJavaWrapper::get_application_context() {
+	if (_get_application_context) {
```
@m4gr3d I might be accidentally re-introducing this here after rebasing everything. Was this one of the methods you moved in your PR?

akien-mga replied:
I don't see any match for `_get_application_context` in the 3.1 branch, nor any commit in the master branch removing it, so I guess this is actually a change you added for this PR specifically.

Whether you actually need/use it, that's for you to say :) @m4gr3d can maybe still comment on whether it's implemented the right way after the recent refactoring.

m4gr3d replied:
`Activity` is a subclass of `Context`, so you can use the already defined `get_activity` method instead for places that need a `Context`.

jahd2602 commented:
@BastiaanOlij Testing with Moto G6 and Pixel 3, and won't detect ARCore ("Failed to start ARCore"). So the app shows the spinning circle but without the camera background, and the camera is 100% static.

When going to play store, it says that the ARCore app is installed.

BastiaanOlij commented:
Did a quick rebase and implemented the changes to the manifest export based on what @m4gr3d did for XR. 

I'll do some more testing tonight/tomorrow but going to see the new spiderman first :)

m4gr3d commented on platform/android/export/export.cpp:
```diff
@@ -1155,6 +1164,7 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 
 	virtual void get_export_options(List<ExportOption> *r_options) {
 
+		r_options->push_back(ExportOption(PropertyInfo(Variant::BOOL, "graphics/arcore"), false));
```
@BastiaanOlij Instead of following this approach, you should add ARCore as another XR mode entry such as the XR mode selection drop-down will show: `Regular`, `Oculus Mobile VR`, `AR Core`.

This also affects the logic to update the manifest file, as you no longer need to add another `meta-data` for ARCore, but you can just reuse the `xr_mode_metadata` one, and set the value accordingly if ARCore is the selected XR mode.

BastiaanOlij replied:
That is a pretty good idea, I'll do that this week

m4gr3d commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -613,6 +615,31 @@ void RasterizerStorageGLES2::texture_allocate(RID p_texture, int p_width, int p_
 	texture->active = true;
 }
 
+void RasterizerStorageGLES2::texture_set_external(RID p_texture, int p_width, int p_height) {
```
@BastiaanOlij Could you add some documentation to highlight the use case for this method here and in the other `rasterizer_*` classes.

BastiaanOlij replied:
Hmmm, not sure if I am the right person to document the rasterizer especially seeing most is being rewritten to Vulkan at the moment.

The external texture is a texture extension in OpenGL that allows a texture to be shared between processes. ARCore requires it because it writes the camera data into that texture from another process (or so I understand).

m4gr3d commented on modules/arcore/arcore_interface.cpp:
```diff
@@ -0,0 +1,696 @@
+		// application must handle these cases at least somewhat gracefully.  See
+		// HelloAR Java sample code for reasonable behavior.
+		ArStatus status = ArCoreApk_requestInstall(env, activity, user_requested_install, &install_status);
+		if (status != AR_SUCCESS) {
```
@BastiaanOlij Another advantage of leveraging the XR Mode interface. We should have some ARCore specific init logic in the Godot.java and GodotView.java which among other things checks that ARCore is installed, and if not, handles the error gracefully.
This will ensure that by the time we get to the native code, all the ARCore install and setup is completed, and we don't need to worry about error cases here.

I'll try to send you a PR to illustrate what I'm describing.

BastiaanOlij replied:
Not sure if that has a whole lot of benefit but I guess it makes sense to have a simple place where permissions are all checked.

m4gr3d replied:
One key advantage to this approach is that in some scenarios, the ARCore install and setup on the device is not fully complete, and the ARCore apis requires the app to fire an intent in order to resolve that. 
This is very hard to do from the native code (e.g: require a lot of plumbing), but very simple to do from the java Android activity layer.

m4gr3d commented on modules/arcore/arcore_interface.h:
```diff
@@ -0,0 +1,117 @@
+/*************************************************************************/
```
@BastiaanOlij What's the reason for ARCore being a module instead of an external plugin?

BastiaanOlij replied:
With OpenVR/Oculus and Oculus mobile we don't have a choice, we rely on external libraries that need to be deployed if these are used that we can't deliver as part of the core application.

With ARCore we're using an API that is part of the OS so the functionality is available, this lowers the barrier of entry a fare amount. It really feels like unnecessary overkill to make this a plugin.

Also when I started it became clear that on iOS you can't have dynamic libraries (I started with ARKit) and at the time GDNative was pretty new, so there were still to many roadblocks. Today, yeah maybe part of this could be placed in a plugin. Whether there is benefit to it? I don't know, like I said I feel its overkill.

m4gr3d replied:
@BastiaanOlij I'm about 3/4 through cleaning up the PR for compatibility with the new Android plugin system and I've got some bad and good news:

- Bad news: This will need to be an external plugin instead of a module. Unlike ARKit, the ARCore API is **not** part of the OS and require a [binary download](https://developers.google.com/ar/develop/c/enable-arcore#dependencies) for access. Asides from the license issues involved with shipping a closed source binary as part of the Godot Engine binary, this will unnecessarily increase the engine binary size, which given the size of the target audience for this feature, does not seem worth it.

- Good news: Looking at the logic here, this should all be doable by leveraging the gdnative-cpp bindings.

BastiaanOlij replied:
@m4gr3d I think the only thing that needs work to move this into a gdnative plugin is that the camera server integration has a few holes. But we can plug those. 

I'm about to get some sleep but let's pick this up through email and see how easily we can transfer the code to a plugin. Seeing most of the relevant code is currently in a module and just an implementation of the ARVRInterface I suspect that most of this will be pretty straight forward especially after all the work we've done for the Quest.

akien-mga replied:
To me the "bad news" of having to make this an external plugin is a great news :)
The legal aspect of having to distribute a proprietary binary was definitely a big blocker for this PR, which was only made as a module as it seemed to be the only/most straightforward option at the time (but this would have likely still prevented actually merging it in core).

So having this moved to a GDNative plugin sounds great. Would it then be an Android plugin (using the new plugin system) that includes the GDNative library?

m4gr3d replied:
@akien-mga Yes it would be an Android plugin with a GDNative library similar to how the Quest plugin is structured.

BastiaanOlij commented:
Added a commit from @m4gr3d with a lot of improvements in it (thanks dude!), I haven't had time to test anything or do any of the finishing touches. Flat out atm :)

BastiaanOlij commented:
Squashed and rebased, everything compiles but I haven't had time to do anything more

dangermousenz commented on platform/android/export/export.cpp:
```diff
@@ -824,17 +836,32 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 							}
 						}
 
 						if (tname == "meta-data" && attrname == "value" && string_table[attr_value] == "xr_mode_metadata_value") {
```
Somehow this line, which isn't showing a change, is throwing the same error on Export as the one below on line 864, that's extra confusing to me, and definitely makes me think I'm doing something wrong

akien-mga replied:
Fixed.

dangermousenz reviewed (commented):
Howdy. I'm totally new to the Godot community, so you're welcome to ignore me. I've given this pull a couple of tries this week and have a few findings. I don't necessarily know if they're real issues or if I just don't know what I'm doing though

If I tweak and comment out the 5 bits that I've mentioned, everything builds and exports just fine! When it runs on my phone, I get the request for access to the camera, which is exciting, and I select allow. However, the camera feed never actually appears to start, and my environment remains default grey

Let me know if you want me to try or test anything else, I'll be happy to help, and I'm super grateful for the work you've put into getting it this far!

akien-mga commented:
Needs another rebase (sorry! :D).

Then I think we should consider it for merging before the 3.2 feature freeze so that we can get tests in 3.2-alpha1, and fix issues until the stable release.

RMKD commented:
So I guess this didn't make the 3.2 freeze? Any updates on when this might be incorporated?

I'm working to test now and running into some issues with compiling the export template on Linux (Manjaro/5.2 kernel) running `scons platform=android target=release_debug -j8`: 
- I had to install and add the path to my specific version of NDK installed from Android Studio: `export ANDROID_NDK_ROOT=$ANDROID_HOME/ndk/20.0.5594570/` to get scons to recognize android as a platform and find the 
- currently getting a file not found error on `platform/android/globals/global_defaults.h`

akien-mga commented:
Needs a rebase after @m4gr3d's changes (note that there are also some buildsystem changes).

BastiaanOlij commented:
Finally rebased this stuff and early testing seem to indicate things are starting to work again :) Many thanks to @akien-mga who's started to play around with this too :)

akien-mga commented on drivers/gles2/shaders/copy.glsl:
```diff
@@ -93,8 +98,10 @@ uniform highp vec4 asym_proj;
+uniform samplerExternalOES source;
 #else
-uniform sampler2D source; // texunit:0
+		uniform sampler2D source; // texunit:0
```
Indentation issue.

akien-mga replied:
Fixed in BastiaanOlij/godot#9.

akien-mga commented on drivers/gles3/shaders/copy.glsl:
```diff
@@ -151,7 +158,7 @@ layout(location = 0) out vec4 frag_color;
 void main() {
 
-	//vec4 color = color_interp;
+//vec4 color = color_interp;
```
Style issue. Old `clang-format` again? :)

akien-mga replied:
Fixed in BastiaanOlij/godot#9.

akien-mga commented on modules/arcore/arcore_interface.h:
```diff
@@ -0,0 +1,116 @@
+/*                           GODOT ENGINE                                */
+/*                    http://www.godotengine.org                         */
+/*************************************************************************/
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
```
Now 2019 :)

akien-mga replied:
Fixed in BastiaanOlij/godot#9.

akien-mga commented on platform/android/export/export.cpp:
```diff
@@ -824,17 +836,32 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
+						if (tname == "uses-sdk" && attrname == "minSdkVersion") {
+							if (xr_mode_index == XR_MODE_ARCORE) {
+								// ARCore requires a min sdk version of 24.
+								string_table.write[attr_value] = "24";
```
This should now be fixed.

akien-mga commented on platform/android/export/export.cpp:
```diff
@@ -824,17 +836,32 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
+							string_table.write[attr_value] = xr_mode_index == XR_MODE_ARCORE ? "true" : "false";
+						}
+
 						if (tname == "meta-data" && attrname == "name" && string_table[attr_value] == "xr_mode_metadata_name") {
```
Should now be fixed.

akien-mga commented on platform/android/SCsub:
```diff
@@ -56,3 +56,16 @@ if lib_arch_dir != '':
+    if 'module_arcore_enabled' in env and env['module_arcore_enabled']:
+        abspath = env.Dir(".").abspath
+        from subprocess import call
+        cmd =  "%s/java/gradlew -b %s/../../thirdparty/arcore/extract-arcore-libs.gradle extractNativeLibraries" % (abspath, abspath)
```
Should be fixed by BastiaanOlij/godot#9 as I now enclose the paths in quotes. Would be great if you can confirm once it's merged/amended into this PR.

BastiaanOlij commented:
Ok @akien-mga , I hope I didn't screw this up but your change is merged, squashed and everything is rebased.

BastiaanOlij commented:
Note that at this point in time the only outstanding work as far as I'm concerned is:
- Code really could use a bit of cleanup
- Plane detection still isn't working for me..

akien-mga commented:
The current state looks quite good to me.

@m4gr3d may want to review the changes I did to the gradle script to download the ARCore SDK.

Now there's still the open question of whether it's legally correct for us to download and distribute the ARCore SDK binaries like this. When you go to the Google download for it, you get https://developers.google.com/ar/develop/downloads

> By downloading the Google ARCore SDK, you agree that the [Google APIs Terms of Service](https://developers.google.com/terms) governs your use thereof.
>
> You must disclose the use of ARCore and how it collects and processes data, prominently in your application, easily accessible to users. You can do this by adding the following text on your main menu or notice screen:
>
> "This application runs on [ARCore](https://play.google.com/store/apps/details?id=com.google.ar.core), which is provided by Google LLC and governed by the [Google Privacy Policy](https://policies.google.com/privacy)."

I don't think we're respecting these terms as is. For other Android API the users have to download the SDK and NDK themselves and agree to their license terms, so that should be fine, but we might need something similar for the ARCore SDK.

Or does the prior acceptance of all licenses from the Android SDK by the user imply that they automatically accept the ARCore SDK ToS too?

m4gr3d commented:
Great effort @BastiaanOlij - I'll take a look this evening!

> The current state looks quite good to me.
> 
> @m4gr3d may want to review the changes I did to the gradle script to download the ARCore SDK.
> 
> Now there's still the open question of whether it's legally correct for us to download and distribute the ARCore SDK binaries like this. When you go to the Google download for it, you get https://developers.google.com/ar/develop/downloads
> 
> > By downloading the Google ARCore SDK, you agree that the [Google APIs Terms of Service](https://developers.google.com/terms) governs your use thereof.
> > You must disclose the use of ARCore and how it collects and processes data, prominently in your application, easily accessible to users. You can do this by adding the following text on your main menu or notice screen:
> > "This application runs on [ARCore](https://play.google.com/store/apps/details?id=com.google.ar.core), which is provided by Google LLC and governed by the [Google Privacy Policy](https://policies.google.com/privacy)."
> 
> I don't think we're respecting these terms as is. For other Android API the users have to download the SDK and NDK themselves and agree to their license terms, so that should be fine, but we might need something similar for the ARCore SDK.
> 
> Or does the prior acceptance of all licenses from the Android SDK by the user imply that they automatically accept the ARCore SDK ToS too?

@akien-mga It doesn't as the ARCore SDK also includes (alongside Android) *iOS*, *Unity*, and *Unreal*.

m4gr3d commented on modules/arcore/arcore_interface.cpp:
```diff
@@ -0,0 +1,649 @@
+		// === ATTENTION!  ATTENTION!  ATTENTION! ===
+		// This method can and will fail in user-facing situations.  Your
+		// application must handle these cases at least somewhat gracefully.  See
+		// HelloAR Java sample code for reasonable behavior.
```
This can be removed since the ArCoreSetupActivity handles the user facing situations.

BastiaanOlij replied:
Cool

m4gr3d commented on modules/arcore/arcore_interface.cpp:
```diff
@@ -0,0 +1,649 @@
+	if (init_status == INITIALISED && ar_session != NULL) {
+		ArStatus status = ArSession_resume(ar_session);
+		if (status != AR_SUCCESS) {
+			// uh, what do we do in this situation?
```
Log and crash?

BastiaanOlij replied:
Sounds like a plan, not sure how to put it into action :)

m4gr3d commented on modules/arcore/arcore_interface.cpp:
```diff
@@ -0,0 +1,649 @@
+			if (is_initialized()) {
+				_pause();
+			}
+			break;
```
New notifications were introduced that are better to detect app resumed and paused state:

- `NOTIFICATION_APP_RESUMED`: Fires when the Android activity is resumed.
- `NOTIFICATION_APP_PAUSED`: Fires when the Android activity is paused.

m4gr3d commented on modules/arcore/arcore_interface.cpp:
```diff
@@ -0,0 +1,649 @@
+		}
+
+		// call resume for the first time ;)
+		_resume();
```
Seems a bit janky to also be invoking `_resume()` here. It might make more sense to have the init logic in `_resume()` here instead.

BastiaanOlij replied:
If I recall correctly the problem is the scenario where you haven't given permission yet because you are running this as your very first ARCore app. Resume is called here but fails, then the user answers the camera permission question and/or installs ARCore and focus returns to our app. 
Resume is then called from NOTIFICATION_WM_FOCUS_IN/NOTIFICATION_APP_RESUMED and now the init succeeds. 

If we put the init in the code here it will fail the first time the app is run, and the user will have to close and re-open the app.

m4gr3d replied:
The `ARCoreSetupActivity` ensures that all the required permissions and/or installs are done before this code is run, so the logic here can assume that by the time it's invoked, everything is good to go!

m4gr3d commented on modules/arcore/arcore_interface.cpp:
```diff
@@ -0,0 +1,649 @@
+		// If we needed to:
+		// - ask permission to use the camera
+		// - prompt to install ARCORE
+		// we'll finish initialising on resume which is called through _notification
```
This is done by the `ARCoreSetupActivity`, so these comments can be removed.

BastiaanOlij commented on drivers/gles2/rasterizer_scene_gles2.cpp:
```diff
@@ -3412,6 +3412,14 @@ void RasterizerSceneGLES2::render_scene(const Transform &p_cam_transform, const
+
+						// TODO: we need to find a better way of doing this
+						storage->shaders.copy.add_custom_define("#extension GL_OES_EGL_image_external : require\n");
+						storage->shaders.copy.set_conditional(CopyShaderGLES2::USE_EXTERNAL_SAMPLER, true);
```
The custom defines system is a bit weird, not sure how big the impact of this is, could be fine as is...

BastiaanOlij commented on modules/arcore/register_types.cpp:
```diff
@@ -0,0 +1,45 @@
+}
+
+void unregister_arcore_types() {
+	// should clean itself up nicely :)
```
Don't know, it was a reminder to myself 2 years ago to figure that out :)
I don't think Godot ever unregisters things but the method structure is there. @akien-mga might now

BastiaanOlij replied:
I'm removing the remark, it doesn't look like any of the other modules ever unregister their classes. I'm now also registering the ARCoreInterface class for completeness sake but it won't have any benefit at this time (see my comment in the main discussion)

BastiaanOlij commented on platform/android/java/lib/build.gradle:
```diff
@@ -2,6 +2,9 @@ apply plugin: 'com.android.library'
     implementation "com.android.support:support-core-utils:28.0.0"
+
+    //ARCore specific, needs to become options
+    implementation "com.google.ar:core:1.10.0"
```
Indeed, I have no sound ideas here which is why the remark is in the code.

BastiaanOlij replied:
@akien-mga we could use your input here. I did originally put all the code in a module so it could in theory be turned off. I'm not sure how we would then exclude these dependencies from the build but it would allow compiling Godot with and without ARCore support.

BastiaanOlij commented:
Ah, i really need to read up on how that works, that wasn’t the case when I
originally wrote it:)

On Thu, 31 Oct 2019 at 4:28 am, Fredia Huya-Kouadio <
notifications@github.com> wrote:

> *@m4gr3d* commented on this pull request.
> ------------------------------
>
> In modules/arcore/arcore_interface.cpp
> <https://github.com/godotengine/godot/pull/26221#discussion_r340757536>:
>
> > +		// call resume for the first time ;)
> +		_resume();
>
> The ARCoreSetupActivity ensures that all the required permissions and/or
> installs are done before this code is run, so the logic here can assume
> that by the time it's invoked, everything is good to go!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/26221?email_source=notifications&email_token=AAO262JIPCHL3QSOEBOYNWTQRG74NA5CNFSM4GZZLFU2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJYOL5Y#discussion_r340757536>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAO262N2Y3BGQFMQRQ76DJDQRG74NANCNFSM4GZZLFUQ>
> .
>
-- 
Kindest regards,

Bastiaan Olij

https://www.facebook.com/bastiaan.olij
https://twitter.com/mux213
https://www.youtube.com/BastiaanOlij
<https://www.youtube.com/channel/UCrbLJYzJjDf2p-vJC011lYw>
https://github.com/BastiaanOlij

BastiaanOlij commented:
One other thing worth discussing, @akien-mga I would specifically like your input here. 

Right now I don't register the ARCoreInterface and ARKitInterface classes, these modules are only compiled in for the mobile platforms so without being part of the other platforms I don't think their documentation is exported anyway. 

I've been thinking about changing the logic so the interfaces are always included but that we have an #IFDEF around the code in each method or ensure they are dummies, that way the classes are available in the documentation and when coding. Alternatively we could have an arcore_interface_dummy.cpp that we compile instead of arcore_interface.cpp on anything but android.

But it feels dirty and ugly to me. 

What is your opinion on this? is there a better way to ensure we have the documentation embedded?

BastiaanOlij commented on platform/android/java/app/AndroidManifest.xml:
```diff
@@ -20,6 +20,11 @@
+    <!--ARCore meta data. This is modified by the exporter based on whether ARCore is enabled-->
+    <uses-feature
+        android:name="xr_mode_uses_camera_name"
+        android:required="true" />
```
I could use your help with that, I still don't 100% follow that code. It also feels that its getting overly messy just to add each feature thing in there. I'm wondering if we can't do something similar to how permissions are implemented. i.e. we have an array that we fill with the feature we require, then at the end we loop through that array and add each feature. I think that would make the code far more readable?

akien-mga commented:
> What is your opinion on this? is there a better way to ensure we have the documentation embedded?

I don't know, that's probably something that we should discuss on IRC with other core developers. That's a problem for various platform-specific APIs, so I think we need to design a clean system to handle this instead of having to define dummy implementations for everything just for the sake of documentation.

dangermousenz commented:
This is building and working perfectly on my Galaxy S8 now, tracking and working nicely.

Is the commit from @m4gr3d all that is needed to render the camera feed now? I'm not comfortable enough with github to attempt to cherry-pick that myself

m4gr3d commented:
@BastiaanOlij We should look into landing this feature for Godot 3.2.3.
I'll take a look at what remains to be done and sync with you.

aaronfranke commented:
@m4gr3d It can't be done in Godot 3.2.3 unless it's done in master first.

m4gr3d commented:
@aaronfranke I think the opposite is likely. Since there are a lot of revisions in master for vulkan support, AR (and VR) support would probably need to be entirely redone.

@BastiaanOlij's work is already (partially) working on the 3.2.x branch however, and needs some cleanup, testing and polishing in order to get it fully ready for release.
Some of that cleanup was dependent on the [Android plugin system re-architecture](https://docs.godotengine.org/en/stable/tutorials/plugins/android/android_plugin.html) which is landing in 3.2.2, so that should unblock it.

m4gr3d commented:
@aaronfranke This PR would probably need to be re-opened against the 3.2 branch.
@BastiaanOlij should be able to shed more light.

BastiaanOlij commented:
Indeed, this is one that has long been planned for 3.2, it's pretty much working for a first iteration and indeed was holding out for the updated plugin architecture. That said it's possible that as a result this should be implemented as a plugin and not as a core feature, I can't oversee the consequences of this yet.

We'll eventually need to add most of this in master but disabled (just like the other AR/VR stuff) but it's far to early to even consider making this work in master.

Problem for me is that I've been flat out the past few months so I haven't had any time to revisit this.  @m4gr3d it would be good if we can spend some time when we're both on discord or something to make some decisions on how we should proceed with this.

zoiba commented:
any chances to have this feature in 3.2.3?

m4gr3d commented:
> any chances to have this feature in 3.2.3?

@zoiba yes, the current plan is to try to land this in the next feature release for Godot.

BastiaanOlij commented:
I've rebased this to the current 3.2 branch. It compiles but it doesn't build the templates because the ARCore dependencies are in wrong. Will be doing some more on this during the week. @m4gr3d could really use your help with pointers on how to change this to the newer build process.

BastiaanOlij commented:
Just recording this here because I've not found time to put finishing touches on this, will get back to this soon I promise. 

Anyway, thanks to Zarasan for this bit of info that should explain why the camera image doesn't work for everyone:

> Hello Bastian, I've been testing different stuff on your arcore pull for the camera texture fix and came across the textureLOD function causing problems
> And found that Mali GPUs dont support textureLOD function
> They only support the texture function
> I changed all of the textureLOD functions to texture only except one and got camera feed on Mali GPUs
> 

@zoiba I hope that also explains the issue you were having.

@m4gr3d I send you an email privately if you can help out with changing the gradlew build, thats the final bit I'm stuck on

veryprofessionaldodo commented:
Hi there! Just checking in, has there been more progress made on this front?

BastiaanOlij commented:
@veryprofessionaldodo unfortunately it's been at a bit of a standstill. I hope to get back into this ASAP and round off the work I was doing last year. 

Note that this PR itself is dead and just kept around until the work to move it all into a plugin is completed. It is superseeded by:
https://github.com/godotengine/godot/pull/41026 for the core changes and
https://github.com/GodotVR/godot_arcore as a plugin.

I got it to compile and make the tracking work again when I got busy on rounding things off at my previous employer.

anthony2i reviewed (approved):
Jailbreak

YuriSizov commented:
> Note that this PR itself is dead

Let's then formally close it. It would need to be redone anyway, even if desired as is.
